### PR TITLE
DRILL-4380: Fix performance regression: in creation of FileSelection …

### PR DESCRIPTION
**Pull request by parthchandra**
…in ParquetFormatPlugin to not set files if metadata cache is available.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java [hunk 1]
```diff
@@ -73,13 +75,18 @@ public String getSelectionRoot() {
...
+    Stopwatch timer = Stopwatch.createStarted();
+
+    if (statuses == null)  {
       final List<FileStatus> newStatuses = Lists.newArrayList();
       for (final String pathStr:files) {
         newStatuses.add(fs.getFileStatus(new Path(pathStr)));
       }
       statuses = newStatuses;
     }
+    logger.info("FileSelection.getStatuses() took {} ms, numFiles: {}",
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java]: DEBUG?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java [hunk 2]
```diff
@@ -118,7 +126,10 @@ public boolean apply(@Nullable FileStatus status) {
       }
     }));
 
-    return create(nonDirectories, null, selectionRoot);
+    final FileSelection fileSel = create(nonDirectories, null, selectionRoot);
+    logger.info("FileSelection.minusDirectories() took {} ms, numFiles: {}",
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java]: same, DEBUG seems more appropriate.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java [hunk 3]
```diff
@@ -233,7 +233,7 @@ private FileSelection expandSelection(DrillFileSystem fs, FileSelection selectio
         // /a/b/c.parquet and the format of the selection root must match that of the file names
         // otherwise downstream operations such as partition pruning can break.
         final Path metaRootPath = Path.getPathWithoutSchemeAndAuthority(metaRootDir.getPath());
-        final FileSelection newSelection = FileSelection.create(null, fileNames, metaRootPath.toString());
+        final FileSelection newSelection = new FileSelection(selection.getStatuses(fs), fileNames, metaRootPath.toString());
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java]: It seems like we keep having issues with misuse of this interface which causes planning regressions. Do you think it makes sense to either change the api or add additional comments to make sure people aren't doing the wrong thing?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java [hunk 4]
```diff
@@ -183,12 +194,16 @@ private static String buildPath(final String[] path, final int folderIndex) {
...
   public static FileSelection create(final DrillFileSystem fs, final String parent, final String path) throws IOException {
+    Stopwatch timer = Stopwatch.createStarted();
     final Path combined = new Path(parent, removeLeadingSlash(path));
     final FileStatus[] statuses = fs.globStatus(combined);
     if (statuses == null) {
       return null;
     }
-    return create(Lists.newArrayList(statuses), null, combined.toUri().toString());
+    final FileSelection fileSel = create(Lists.newArrayList(statuses), null, combined.toUri().toString());
+    logger.info("FileSelection.create() took {} ms ", timer.elapsed(TimeUnit.MILLISECONDS));
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java]: INFO => DEBUG

**Comment by jacques-n** [comment]
Other than INFO => DEBUG, +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java (hunk 2 above)
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java]: minusDirectories() ?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java (hunk 3 above)
Comment (hnfgns) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java]: Whole point of making this c'tor non public was to centralize creation via FileSelection.create(...). Looks like we need more explicit comments over here. For this patch, a public c'tor seems not required as well. FileSelection.create(selections, null, root) should do the trick.

Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java]: Agreed. Hanifi made the change to the api initially with exactly that in mind. This patch reverts that partially. He's logging a new Jira to fix the api and document usage.

Comment (hnfgns) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java]: Filed DRILL-4381. Thanks.

**Pull request closed without merging**